src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java

```diff
@@ -0,0 +1,681 @@
+
+    @Argument(doc="Format use to name count column headers. It can be any string where the following special  " +
+            "sequences would be substituted: %^ -> default name given the groupBy option, %s -> sample-id, %r -> read-group-id, %c -> cohort name", optional=true,
+       fullName="countColumnName",shortName="countColumnName")
+    protected String countColumnName = "%^";
+
+    private PrintWriter outputWriter;
+
+    private PrintWriter summaryPerColumnOutputWriter;
+
+    private PrintWriter summaryPerRowOutputWriter;
+
+    private SampleDB sampleDB;
+
+    private int groupCount = 1;
+
+    private static final Logger logger = LogManager.getLogger(ExomeReadCounts.class);
+
+    private ExonDB<GenomeLoc> exonDB;
+
+    private Counts counts;
+
+    private GenomeLocParser genomeLocFactory;
+
+    @Override
+    public void onTraversalStart() {
+        super.onTraversalStart();
+        final SAMSequenceDictionary sequenceDictionary = getBestAvailableSequenceDictionary();
+        if (sequenceDictionary == null)
+            throw new UserException("missing reference sequence dictionary, as no input contains such a dictionary... please consider to provide fail-over dictionary using " + 
+                                    "a reference (e.g. --reference myref.fasta)");
+        if (!intervalArgumentCollection.intervalsSpecified())
+            throw new UserException("you must indicate the set of exon as input intervals");
+
+        genomeLocFactory = new GenomeLocParser(sequenceDictionary);
+        sampleDB = new SampleDB(getHeaderForReads());
+        logger.log(Level.INFO,"Reading exons locations from intervals...");
+        final List<GenomeLoc> genomeLocs =
+                fromSimpleIntervalsToGenomeLocs(genomeLocFactory,intervalArgumentCollection.getIntervals(sequenceDictionary));
+        exonDB = new IntervalBackedExonDB(genomeLocs);
+        logger.log(Level.INFO,String.format("   Found %d exons.", exonDB.exonCount()));
+        logger.log(Level.INFO,"...done.");
+        counts = createCounts();
+        outputWriter = openOutputWriter(output);
+        if (summaryPerColumnOutput != null)
+            summaryPerColumnOutputWriter = openOutputWriter(summaryPerColumnOutput);
+        if (summaryPerRowOutput != null)
+            summaryPerRowOutputWriter = openOutputWriter(summaryPerRowOutput);
+
+        logger.log(Level.INFO,"Calculating exon depths...");
+    }
+
+    private List<GenomeLoc> fromSimpleIntervalsToGenomeLocs(final GenomeLocParser genomeLocFactory,
+                                                            final List<SimpleInterval> intervals) {
+        if (intervals == null || intervals.isEmpty())
+            return Collections.emptyList();
+        else if (intervals.size() == 1) {
+            final Locatable locatable = intervals.get(0);
+            return Collections.singletonList(genomeLocFactory.createGenomeLoc(locatable.getContig(),locatable.getStart(),locatable.getEnd()));
```

Comment:
This view ("premature optimization is always wrong. The only OK optimization is that backed by profile data.") is a bit too extreme, I think. Even before profiling, we should not be intentionally doing egregiously inefficient things. For example, it's good practice to use `StringBuilder` when incrementally building up a large String instead of string concatenation. It's only when the optimization falls into the category of a micro-optimization that seriously harms code readability that we should start screaming "premature optimization".